### ISSUE TITLE
Transparent PNG are not always displayed properly in dark mode

### ISSUE THREAD

**Issue by kelson42** [openzim/mwoffliner#1026]
In dark mode, the bakground is black. If the PNG is a black design on transparent background, then nothing will be visible for the user. This is what happen with this article http://library.kiwix.org/wikipedia_fr_mathematics_maxi/A/Parabole on Android for example.

**Comment by Jaifroid** [comment]
But how can mwoffliner help? The scraper should always scrape the file in the default mode, which is light. It would then be up to the OS or the reader to invert the image if it is either an SVG or a PNG based on an SVG. In Kiwix JS Windows, I have a rule (or two!) in the dark mode stylesheet that inverts the image in this situation, ensuring that equations and other images are visible in dark mode. Most dark modes avoid inverting the image, because it looks horrible to the user, but unfortunately some images must be inverted and I think the decision has to be made by the reader, not the scraper.

**Comment by kelson42** [comment]
@Jaifroid By surrending the IMG in a `<div>` of the exact same size and with a white (or same colour like the background) background. In light mode, no difference, and in darmode, we will have a square with white bacgkround (so no transparency here).

**Comment by Jaifroid** [comment]
@kelson42 IMHO this would be a dirty implementation because it would hard-code a light background into the source, negate the transparency in the image, and defeat the purpose of a dark mode (to avoid being blinded by white background at night, for example). It would be better to mark these images as needing special treatment by the reader (in the same way as SVG equations as stated in https://github.com/kiwix/kiwix-android/issues/1649 ) but not hard-code any specific colour background into them.

**Comment by kelson42** [comment]
@Jaifroid I'm against any kind of hack at the reader level for obvious. As long as we don't have JS API, what I have proposed seems to be the best compromise to me between short and long term goals.

**Comment by Jaifroid** [comment]
@kelson42 OK, but please be very careful not to surround equations (which are SVG) with a hard-coded white background. These are inline and are read as part of the text of the page. I give an example of equations rendered in dark mode below (from a ZIM without the MathML). Just imagine how horrible this would look if each of the equations on this page had a white background...

![image](https://user-images.githubusercontent.com/4304337/71160491-91a4c480-223f-11ea-9976-2b22787f2933.png)

**Comment by kelson42** [comment]
@Jaifroid This is this problem or you can have content which is black on black... what is the best to your opinion?

**Comment by Jaifroid** [comment]
In light mode, equations need to be presented as black text against the light background, and in dark mode they need to be presented as white text against the dark background. As @macgills commented in https://github.com/kiwix/kiwix-android/issues/1649#issuecomment-566527524 , it's just a simple stylesheet entry to achieve this, not a hack. A properly implemented dark mode is a popular request of users (comments on Kiwix JS Windows app repeatedly appreciate the dark mode).

**Comment by kelson42** [comment]
@Jaifroid This is a hack because the reader should not have any content specific treatment. It should not because by doing so you create a dependence between the reader and the type of content which is afterward hard to manage. If tomorrow we fix the problem in the HTML, then this hack will start to behave badly, so you will create a new problem just by having solved the problem properly.

**Comment by macgills** [comment]
equally a grey background can provide contrast on white/black. Adding a background to a transparent image is always going to carry a risk of obscuring something as we cannot guarantee what the content is

**Comment by kelson42** [comment]
And just to be clear. I perfectly understand that there is not good solution for now in the way how we deal with the dark-mode problem The solution should be to have a JS function within the ZIM files doing this work and then called by the reader. This is what we call the JS API. Not a huge deal, but we are short on resources for now in JS.

**Comment by Jaifroid** [comment]
OK, I tend to be quite user-orientated and not very purist about workarounds. I do fear that if you surround all SVGs and transparent PNGs with a white background, it's going to make a mess of dark mode for mathematics articles. Dark mode is a very popular option now that it's been embraced fully by Android and iOS (Windows Mobile had it since 2007, he he 🙂), and I think we need to do it properly in all readers.

> This is the best Kiwix client for Windows 10. It looks awesome, **I like the dark mode too**.

> I like it more than desktop app, this is lighter consumes 100MB while the other 300MB, is more optimized, has better search result and **fabulous dark mode**

**Comment by kelson42** [comment]
@Jaifroid Yes dark mode is really popular and not only on Windows and I appreciate that everything is done on Kiwix-JS to provide the best UX. But, I have to care to many other ports and secure an overall healthy solution architecture. It has to be sustainable, ie. robust whatever the content and not needing a lot of complex maintenance.

**Comment by danielzgtg** [comment]
A checkerboard background is another option and would be less likely to obscure something important.

**Comment by Mitia99** [comment]
I also hope that this issue will be fixed.

**Comment by MananJethwani** [comment]
@Jaifroid @kelson42 for transparent images we can add a white border stroke to the image in a css file that could do the job.

**Comment by Mitia99** [comment]
Hope that it can be considered as priority cause after 4 years, I noticed that it is not yet fixed in Kiwix (Android)
![AddText_07-23-09.24.42_edit_3220058479716.jpg](https://github.com/kiwix/kiwix-android/assets/56766778/931e9e7c-d58f-476a-ba10-a982856bd462)

**Comment by Jaifroid** [comment]
As explained above, the dark mode stylesheet in the reader needs a rule that adds a light-gray background to SVG images that are not equations (even if the SVG has been converted to WebP). It's how it's done in the PWA and in the browser extension (see below). It can be done with just three stylesheet rules you can see in the [Kiwix JS source code](https://github.com/kiwix/kiwix-js/blob/main/www/css/kiwixJS_mwInvert.css).

![image](https://github.com/openzim/mwoffliner/assets/4304337/96973e46-5b0d-4598-84db-2da7ea51f8b7)

![image](https://github.com/openzim/mwoffliner/assets/4304337/037dade1-cdcd-4ea2-9437-5119e68b7527)

**Comment by MohitMaliFtechiz** [comment]
> As explained above, the dark mode stylesheet in the reader needs a rule that adds a light-gray background to SVG images that are not equations (even if the SVG has been converted to WebP). It's how it's done in the PWA and in the browser extension (see below). It can be done with just three stylesheet rules you can see in the [Kiwix JS source code](https://github.com/kiwix/kiwix-js/blob/main/www/css/kiwixJS_mwInvert.css).

@Jaifroid We have an issue on Kiwix-android https://github.com/kiwix/kiwix-android/issues/4234, and this issue also exists on the kiwix js extension and  https://pwa.kiwix.org/.

`Kiwix js`  browser extension if I use "Dark (generic invert)" from "Display settings".

![Image](https://github.com/user-attachments/assets/43153fc0-79a0-4271-80ac-f7c9e9e514e6)

![Image](https://github.com/user-attachments/assets/d4bd75aa-2ca6-4b4c-b0b6-8c8d1fc1c87c)

**pwa.kiwix.org**

![Image](https://github.com/user-attachments/assets/60be9f7d-564a-45e2-8808-299cf3ca4cd5)

![Image](https://github.com/user-attachments/assets/cee266ba-5fd1-4626-b66d-46971c14ea73)

**Comment by benoit74** [comment]
I don't get why we are trying to make dark mode work here. 

I've heard that so far the recommendation is to not handle dark mode in scrapers / ZIMs because we are too far from being able to make it work properly.

So I'm a bit puzzled by the idea of finding how to invert the image.

For me, the position so far is that readers should be aware that ZIMs cannot go dark, and as such they should override system preference around dark mode by choosing "light only". Or scrapers should set a high-level CSS directive to force light mode.

Do I miss something? Or is it just that since 2019 our maturity around our capability to use dark mode has changed?

For instance, why https://library.kiwix.org/viewer#wikipedia_en_mathematics_maxi_2025-03/A/Parabola is white background despite my browser being in dark mode?

**Comment by kelson42** [comment]
With https://github.com/openzim/overview/issues/53 it is now clear that technical details about how to display in a dark mode is the responsability of the ZIM file itself or better said: of the scraper.

As a consequence, the only responsibility of the reader is to tell the HTML render element what theme (light/dark).

For the moment there is no dark theme support in MWoffliner, but this is the role of implementing #1136
